src/main/drivers/flash_w25m.c

```diff
@@ -0,0 +1,242 @@
+        fdevice->geometry.sectorSize = 0;
+        fdevice->geometry.totalSize = 0;
+        return false;
+    }
+
+    fdevice->geometry.sectors = dieDevice[0].geometry.sectors;
+    fdevice->geometry.sectorSize = dieDevice[0].geometry.sectorSize;
+    fdevice->geometry.pagesPerSector = dieDevice[0].geometry.pagesPerSector;
+    fdevice->geometry.pageSize = dieDevice[0].geometry.pageSize;
+    dieSize = dieDevice[0].geometry.totalSize;
+    fdevice->geometry.totalSize = dieSize * dieCount;
+    fdevice->vTable = &w25m_vTable;
+
+    return true;
+}
+
+void w25m_eraseSector(flashDevice_t *fdevice, uint32_t address)
+{
+    int dieNumber = address / dieSize;
+
+    w25m_dieSelect(fdevice->busdev, dieNumber);
+
+    dieDevice[dieNumber].vTable->eraseSector(&dieDevice[dieNumber], address % dieSize);
+}
+
+void w25m_eraseCompletely(flashDevice_t *fdevice)
+{
+    for (int dieNumber = 0 ; dieNumber < dieCount ; dieNumber++) {
+        w25m_dieSelect(fdevice->busdev, dieNumber);
+        dieDevice[dieNumber].vTable->eraseCompletely(&dieDevice[dieNumber]);
+    }
+}
+
+static uint32_t currentWriteAddress;
+static int currentWriteDie;
+
+void w25m_pageProgramBegin(flashDevice_t *fdevice, uint32_t address)
+{
+    UNUSED(fdevice);
+
+    currentWriteDie = address / dieSize;
+    w25m_dieSelect(fdevice->busdev, currentWriteDie);
+    currentWriteAddress = address % dieSize;
+    dieDevice[currentWriteDie].vTable->pageProgramBegin(&dieDevice[currentWriteDie], currentWriteAddress);
+}
+
+void w25m_pageProgramContinue(flashDevice_t *fdevice, const uint8_t *data, int length)
+{
+    UNUSED(fdevice);
+
+    dieDevice[currentWriteDie].vTable->pageProgramContinue(&dieDevice[currentWriteDie], data, length);
+}
+
+void w25m_pageProgramFinish(flashDevice_t *fdevice)
+{
+    UNUSED(fdevice);
+
+    dieDevice[currentWriteDie].vTable->pageProgramFinish(&dieDevice[currentWriteDie]);
+}
+
+void w25m_pageProgram(flashDevice_t *fdevice, uint32_t address, const uint8_t *data, int length)
+{
+    w25m_pageProgramBegin(fdevice, address);
+
+    w25m_pageProgramContinue(fdevice, data, length);
+
+    w25m_pageProgramFinish(fdevice);
+}
+
+int w25m_readBytes(flashDevice_t *fdevice, uint32_t address, uint8_t *buffer, int length)
+{
+    int rlen; // remaining length
+    int tlen; // transfer length for a round
+    int rbytes;
+
+    // Divide a read that spans multiple dies into two.
+    // The loop is executed twice at the most for decent 'length'.
+
+    for (rlen = length; rlen; rlen -= tlen) {
+        int dieNumber = address / dieSize;
+        uint32_t dieAddress = address % dieSize;
+
+        if (dieAddress + rlen > dieSize) {
+            tlen = dieSize - dieAddress;
+        } else { 
+            tlen = rlen;
+        }
+
+        w25m_dieSelect(fdevice->busdev, dieNumber);
+
+        rbytes = dieDevice[dieNumber].vTable->readBytes(&dieDevice[dieNumber], dieAddress, buffer, tlen);
+
+        if (!rbytes) {
+            debug[0]++;
```

Comment:
This will have to be changed to use `DEBUG_SET()` or removed, or else it will break other debug modes.

Reply:
Oops...